File: packages/python/src/opensms/resources/contact_groups.py
"""The ``contact_groups`` resource."""

from __future__ import annotations

from typing import Dict, List, Optional

from .._serialize import path_id, prune
from ..models import Batch, ContactGroup, Page
from ._base import Resource


class ContactGroups(Resource):
    """Groups of contacts. Accessed as ``client.contact_groups``."""

    def list(self, *, limit: Optional[int] = None, cursor: Optional[str] = None) -> Page[ContactGroup]:
        """List contact groups."""
        return self._page(
            self._t.request("GET", "/v1/contact-groups", query={"limit": limit, "cursor": cursor})
        )

    def create(
        self,
        *,
        name: str,
        contact_ids: Optional[List[str]] = None,
        idempotency_key: Optional[str] = None,
    ) -> ContactGroup:
        """Create a group."""
        body = prune({"name": name, "contact_ids": contact_ids})
        return self._t.request(
            "POST", "/v1/contact-groups", json_body=body, idempotent=True, idempotency_key=idempotency_key
        )

    def get(self, id: str) -> ContactGroup:
        """Fetch a group."""
        return self._t.request("GET", f"/v1/contact-groups/{path_id(id)}")

    def update(
        self,
        id: str,
        *,
        name: Optional[str] = None,
        contact_ids: Optional[List[str]] = None,
    ) -> ContactGroup:
        """Rename a group and/or replace its members."""
        body = prune({"name": name, "contact_ids": contact_ids})
        return self._t.request("PATCH", f"/v1/contact-groups/{path_id(id)}", json_body=body)

    def delete(self, id: str) -> None:
        """Delete a group (contacts are kept)."""
        self._t.request("DELETE", f"/v1/contact-groups/{path_id(id)}")

    def send(
        self,
        id: str,
        *,
        text: Optional[str] = None,
        template_id: Optional[str] = None,
        variables: Optional[Dict[str, str]] = None,
        sender_id: Optional[str] = None,
        traffic_type: Optional[str] = None,
        callback_url: Optional[str] = None,
        idempotency_key: Optional[str] = None,
    ) -> Batch:
        """Send ``text`` or a ``template_id`` to every member. Returns a
        running Batch."""
        body = prune(
            {
                "text": text,
                "template_id": template_id,
                "variables": variables,
                "sender_id": sender_id,
                "traffic_type": traffic_type,
                "callback_url": callback_url,
            }
        )
        return self._t.request(
            "POST",
            f"/v1/contact-groups/{path_id(id)}/send",
            json_body=body,
            idempotent=True,
            idempotency_key=idempotency_key,
        )

File: packages/ruby/lib/opensms/resources/contact_groups.rb
# frozen_string_literal: true

require_relative "base"

module Opensms
  module Resources
    # The +contact_groups+ resource. Accessed as +client.contact_groups+.
    class ContactGroups < Base
      FIELDS = %i[name contact_ids].freeze
      SEND_FIELDS = %i[text template_id variables sender_id traffic_type callback_url].freeze

      # GET /v1/contact-groups -> Page<Group>.
      # @param params [Hash] :limit, :cursor
      def list(params = nil, **kwargs)
        http_page("/v1/contact-groups", Models.build(merge(params, kwargs), %i[limit cursor]))
      end

      # POST /v1/contact-groups -> 201 Group.
      # @param params [Hash] :name (required), :contact_ids
      def create(params = nil, idempotency_key: nil, **kwargs)
        body = Models.build(merge(params, kwargs), FIELDS, required: %i[name])
        @transport.request(:post, "/v1/contact-groups", body: body, idempotency_key: idem(idempotency_key))
      end

      # GET /v1/contact-groups/{id} -> Group.
      def get(id)
        http_get(path("/v1/contact-groups", id!(id)))
      end

      # PATCH /v1/contact-groups/{id} -> Group.
      # @param params [Hash] :name, :contact_ids
      def update(id, params = nil, **kwargs)
        body = Models.build(merge(params, kwargs), FIELDS)
        @transport.request(:patch, path("/v1/contact-groups", id!(id)), body: body)
      end

      # DELETE /v1/contact-groups/{id} -> 204.
      # @return [nil]
      def delete(id)
        http_delete(path("/v1/contact-groups", id!(id)))
      end

      # Send to every contact in the group. POST /v1/contact-groups/{id}/send
      # -> Batch (already "running").
      #
      # @param params [Hash] :text or :template_id, :variables, :sender_id,
      #   :traffic_type, :callback_url
      def send(id, params = nil, idempotency_key: nil, **kwargs)
        body = Models.build(merge(params, kwargs), SEND_FIELDS)
        @transport.request(:post, path("/v1/contact-groups", id!(id), :send), body: body,
                                                                             idempotency_key: idem(idempotency_key))
      end
    end
  end
end
